### PR TITLE
Give the AI a "Reset All APCs" command

### PR DESCRIPTION
**Pull request by FlameArrow57**
[FEATURE][BALANCE]

## About the PR 
This PR gives the AI a new command, "Reset All APCs." What it does is reset all APCs on the station that do not have AI control disabled to their default settings, with some settings changed dependent on if the APC is shorted or not.

## Why's this needed? 
It's useful for if the AI is rogue, and the AI or cyborgs turn off a lot of power, and then the AI is unrogued. Similar to un-electrify all doors and unbolt all doors, but for APCs.

## Changelog
```changelog
(u)FlameArrow57
(*)The AI now has a "Reset All APCs" command, which will attempt to reset all APCs to their default settings.
(+)Ion storms may now disable AI control on some APCs.
```

### CODE REVIEW

File: code/modules/power/apc.dm [hunk 1]
```diff
@@ -1548,3 +1548,20 @@ var/zapLimiter = 0
 /obj/machinery/power/apc/powered()
 	//Always powered
 	return 1
+
+/obj/machinery/power/apc/proc/is_not_default()
+	return operating != 1 || chargemode != 1 || (!shorted && (equipment != 3 || lighting != 3 || environ != 3)) || coverlocked != 1
```
Comment (Zonespace27) [inline code/modules/power/apc.dm:1553]: if `operating`, `chargemode` and others can only be 1/0, turn them to TRUE/FALSE

File: code/mob/living/silicon/ai.dm [hunk 2]
```diff
@@ -1728,6 +1729,28 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
...
+	set category = "AI Commands"
+	set name = "Reset All APCs"
+	set desc = "Resets all APCs on the station."
+	var/count = 0
+
+	var/mob/message_mob = src.get_message_mob()
+	if (!src || !message_mob.client || isdead(src))
+		return
+
+	if(alert("Are you sure?",,"Yes","No") == "Yes")
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:1742]: tgui alert

File: code/mob/living/silicon/ai.dm [hunk 3]
```diff
@@ -1728,6 +1729,28 @@ var/list/ai_emotions = list("Happy" = "ai_happy",\
...
+		for_by_tcl(P, /obj/machinery/power/apc)
+			if (P.z == 1 && !(P.status & BROKEN) && !P.aidisabled && P.is_not_default())
```
Comment (ZeWaka) [inline code/mob/living/silicon/ai.dm:1744]: station z level

File: code/modules/power/apc.dm [hunk 4]
```diff
@@ -1548,3 +1548,18 @@ var/zapLimiter = 0
 /obj/machinery/power/apc/powered()
 	//Always powered
 	return 1
+
+/obj/machinery/power/apc/proc/is_not_default()
+	return !operating || !chargemode || (!shorted && (equipment != 3 || lighting != 3 || environ != 3)) || !coverlocked
```
Comment (ZeWaka) [inline code/modules/power/apc.dm:1553]: use initial() instead

**Review by ZeWaka — APPROVED** [review]

**Comment by Tarmunora** [comment]
Hm. Might want to make ion storms mess with AI control on APCs, since I don't think that trivializing the APC-fuckery from ion storms was an intended side-effect

**Comment by FlameArrow57** [comment]
> Hm. Might want to make ion storms mess with AI control on APCs, since I don't think that trivializing the APC-fuckery from ion storms was an intended side-effect
> 
> Though I suppose the AI can trivialize the door parts of it currently :lazy:

Sure. It looks like there's already some code that messes up 20 random (it seems) APCs on the station when an ion storm occurs.

image

Should I just add another case for turning off ai control, or maybe just make it so that all 20 have ai control disabled, or something else?

**Comment by Tarmunora** [comment]
> Should I just add another case for turning off ai control, or maybe just make it so that all 20 have ai control disabled, or something else?

either for all of them, or a 50% chance to do disable AI control I think.

**Comment by FlameArrow57** [comment]
I updated the changelog to show the ion storm APC change in case it's needed.

**Review by Tarmunora — APPROVED** [review]

**Pull request merged**